src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java

```diff
@@ -670,6 +702,73 @@ private int updateStopTimesForPatternStop(ObjectNode patternStop, int previousTr
     }
 
+    /**
+     * Normalizes all stop times' arrivals and departures for an ordered set of pattern stops. This set can be the full
+     * set of stops for a pattern or just a subset. Typical usage for this method would be to overwrite the arrival and
+     * departure times for existing trips after a pattern stop has been added or inserted into a pattern or if a
+     * pattern stop's default travel or dwell time were updated and the stop times need to reflect this update.
+     * @param patternStops list of pattern stops for which to update stop times (ordered by increasing stop_sequence)
+     * @param serviceIdFilters service_id values to filter trips on
+     * @throws SQLException
+     */
+    private int updateStopTimesForPatternStops(List<PatternStop> patternStops, Set<String> serviceIdFilters) throws SQLException {
+        PatternStop firstPatternStop = patternStops.iterator().next();
+        int firstStopSequence = firstPatternStop.stop_sequence;
+        // Prepare SQL query to determine the time that should form the basis for adding the travel time values.
+        int previousStopSequence = firstStopSequence > 0 ? firstStopSequence - 1 : 0;
+        String timeField = firstStopSequence > 0 ? "departure_time" : "arrival_time";
+        String getFirstTravelTimeSql = String.format(
+            "select t.trip_id, %s from %s.stop_times st, %s.trips t where stop_sequence = ? " +
+                "and t.pattern_id = ? " +
+                "and t.trip_id = st.trip_id",
+            timeField,
+            tablePrefix,
+            tablePrefix
+        );
+        if (serviceIdFilters != null && serviceIdFilters.size() > 0) {
+            getFirstTravelTimeSql += String.format(
+                " and t.service_id in (%s)",
+                Collections.nCopies(serviceIdFilters.size(), "?")
+            );
+        }
+        PreparedStatement statement = connection.prepareStatement(getFirstTravelTimeSql);
+        statement.setInt(1, previousStopSequence);
+        statement.setString(2, firstPatternStop.pattern_id);
+        LOG.info(statement.toString());
+        ResultSet resultSet = statement.executeQuery();
+        Map<String, Integer> timesForTripIds = new HashMap<>();
+        while (resultSet.next()) {
+            timesForTripIds.put(resultSet.getString(1), resultSet.getInt(2));
+        }
+        // Update stop times for individual trips with normalized travel times.
+        String updateTravelTimeSql = String.format(
+            "update %s.stop_times set arrival_time = ?, departure_time = ? where trip_id = ? and stop_sequence = ?",
+            tablePrefix
+        );
+        PreparedStatement updateStopTimeStatement = connection.prepareStatement(updateTravelTimeSql);
+        LOG.info(updateStopTimeStatement.toString());
+        final BatchTracker stopTimesTracker = new BatchTracker("stop_times", updateStopTimeStatement);
+        for (String tripId : timesForTripIds.keySet()) {
```

Comment:
Could this for loop be avoided by doing the update statement using a join to find all applicable trip IDs?

Reply:
Oh, hmm, I see why there are 2 loops. It could still probably be combined into one huge messy SQL query, but I suppose this is fine for now.

Reply:
🤔... is `one huge messy SQL query` a desirable thing? I don't think it necessarily is.

Reply:
It's only desirable if there are considerable performance benefits. However, that would entail much more investigation in running jmeter tests and whatnot so yeah, I'm fine with keeping this as is until users ever start complaining about the performance of this.

Reply:
And even if there are considerable performance benefits, it's only desirable if the code is factored to remain readable, and that performance gain is actually needed and a significant fraction of total runtime. Otherwise in my opinion the best thing to do is add a comment informing future readers of the alternative approach.